File: implementations/python/src/reifinator/resolution.py
"""Filename placeholder resolution and generation tracking."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Protocol


class PlaceholderResolver(Protocol):
    """Protocol for placeholder resolution strategies."""

    def resolve(self, filename: str, context: dict[str, Any]) -> FilenameResolution:
        """Resolve placeholders in a filename against the given context."""
        ...


@dataclass
class FilenameResolution:
    """Result of attempting to resolve a filename with embedded expressions."""

    original: str
    resolved: str | None  # None if resolution failed
    expressions: list[str]  # All expressions found in the filename

    @property
    def success(self) -> bool:
        return self.resolved is not None

    @property
    def has_expressions(self) -> bool:
        return len(self.expressions) > 0

    @property
    def was_substituted(self) -> bool:
        return self.success and self.resolved != self.original


class BracketResolver:
    """Resolves [expression] placeholders using dot-notation property access.

    This is the default resolver. Expressions like [entity.name] are resolved
    by walking the context dict (and object attributes as fallback).
    """

    PATTERN = re.compile(r"\[([\w.]+)\]")

    def resolve(self, filename: str, context: dict[str, Any]) -> FilenameResolution:
        resolution_failed = False
        expressions_found: list[str] = []

        def replace(match: re.Match) -> str:
            nonlocal resolution_failed
            expr = match.group(1)
            expressions_found.append(expr)
            parts = expr.split(".")
            value: Any = context
            try:
                for part in parts:
                    if isinstance(value, dict):
                        if part not in value:
                            resolution_failed = True
                            return match.group(0)
                        value = value[part]
                    else:
                        value = getattr(value, part)
                return str(value)
            except (AttributeError, KeyError):
                resolution_failed = True
                return match.group(0)

        result = self.PATTERN.sub(replace, filename)
        return FilenameResolution(
            original=filename,
            resolved=None if resolution_failed else result,
            expressions=expressions_found,
        )


@dataclass
class GenerationTracker:
    """Tracks template items with expressions and whether they were resolved.

    After a generation run, any item that contained expressions but was never
    resolved indicates a likely error (typo, missing context variable).
    """

    expression_items: dict[Path, list[str]] = field(default_factory=dict)
    resolved_items: set[Path] = field(default_factory=set)

    def register_expression_item(self, item_path: Path, expressions: list[str]) -> None:
        if item_path not in self.expression_items:
            self.expression_items[item_path] = expressions

    def mark_resolved(self, item_path: Path) -> None:
        self.resolved_items.add(item_path)

    def get_unresolved_items(self) -> list[tuple[Path, list[str]]]:
        return [
            (path, exprs)
            for path, exprs in self.expression_items.items()
            if path not in self.resolved_items
        ]

    def raise_if_unresolved(self) -> None:
        unresolved = self.get_unresolved_items()
        if unresolved:
            raise UnresolvedExpressionsError(unresolved)


class UnresolvedExpressionsError(Exception):
    """Raised when template items with expressions could not be resolved."""

    def __init__(self, unresolved_items: list[tuple[Path, list[str]]]):
        self.unresolved_items = unresolved_items
        items_to_show = unresolved_items[:5]
        lines = ["The following template items have expressions that could not be resolved:"]
        for item_path, expressions in items_to_show:
            expr_str = ", ".join(f"[{e}]" for e in expressions)
            lines.append(f"  - {item_path}: {expr_str}")
        if len(unresolved_items) > 5:
            lines.append(f"  ... and {len(unresolved_items) - 5} more")
        super().__init__("\n".join(lines))

File: implementations/typescript/src/resolution.ts
/**
 * Filename placeholder resolution and generation tracking.
 */

export interface FilenameResolution {
  original: string;
  resolved: string | null; // null if resolution failed
  expressions: string[]; // all expressions found in the filename
  success: boolean;
  wasSubstituted: boolean;
}

export interface PlaceholderResolver {
  resolve(filename: string, context: Record<string, unknown>): FilenameResolution;
}

/**
 * Resolves [expression] placeholders using dot-notation property access.
 */
export class BracketResolver implements PlaceholderResolver {
  private static readonly PATTERN = /\[([\w.]+)\]/g;

  resolve(
    filename: string,
    context: Record<string, unknown>,
  ): FilenameResolution {
    let resolutionFailed = false;
    const expressionsFound: string[] = [];

    const result = filename.replace(
      BracketResolver.PATTERN,
      (match, expr: string) => {
        expressionsFound.push(expr);
        const parts = expr.split(".");
        let value: unknown = context;

        try {
          for (const part of parts) {
            if (value === null || value === undefined) {
              resolutionFailed = true;
              return match;
            }
            if (typeof value === "object" && part in (value as object)) {
              value = (value as Record<string, unknown>)[part];
            } else {
              resolutionFailed = true;
              return match;
            }
          }
          return String(value);
        } catch {
          resolutionFailed = true;
          return match;
        }
      },
    );

    const resolved = resolutionFailed ? null : result;
    return {
      original: filename,
      resolved,
      expressions: expressionsFound,
      success: !resolutionFailed,
      wasSubstituted: !resolutionFailed && result !== filename,
    };
  }
}

/**
 * Tracks template items with expressions and whether they were resolved.
 */
export class GenerationTracker {
  private expressionItems = new Map<string, string[]>();
  private resolvedItems = new Set<string>();

  registerExpressionItem(itemPath: string, expressions: string[]): void {
    if (!this.expressionItems.has(itemPath)) {
      this.expressionItems.set(itemPath, expressions);
    }
  }

  markResolved(itemPath: string): void {
    this.resolvedItems.add(itemPath);
  }

  getUnresolvedItems(): Array<[string, string[]]> {
    const unresolved: Array<[string, string[]]> = [];
    for (const [path, exprs] of this.expressionItems) {
      if (!this.resolvedItems.has(path)) {
        unresolved.push([path, exprs]);
      }
    }
    return unresolved;
  }

  raiseIfUnresolved(): void {
    const unresolved = this.getUnresolvedItems();
    if (unresolved.length > 0) {
      throw new UnresolvedExpressionsError(unresolved);
    }
  }
}

export class UnresolvedExpressionsError extends Error {
  public readonly unresolvedItems: Array<[string, string[]]>;

  constructor(unresolvedItems: Array<[string, string[]]>) {
    const itemsToShow = unresolvedItems.slice(0, 5);
    const lines = [
      "The following template items have expressions that could not be resolved:",
    ];
    for (const [itemPath, expressions] of itemsToShow) {
      const exprStr = expressions.map((e) => `[${e}]`).join(", ");
      lines.push(`  - ${itemPath}: ${exprStr}`);
    }
    if (unresolvedItems.length > 5) {
      lines.push(`  ... and ${unresolvedItems.length - 5} more`);
    }
    super(lines.join("\n"));
    this.name = "UnresolvedExpressionsError";
    this.unresolvedItems = unresolvedItems;
  }
}
